server/src/main/java/io/spine/server/storage/QueryConverter.java

```diff
@@ -0,0 +1,309 @@
+import static io.spine.util.Exceptions.newIllegalArgumentException;
+import static io.spine.util.Exceptions.newIllegalStateException;
+import static java.lang.String.join;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Converts the queries defined in Protobuf into the language of {@code io.spine.query} package.
+ */
+@Internal
+public final class QueryConverter {
+
+    private QueryConverter() {
+    }
+
+    /**
+     * Converts the Protobuf-based filters and the response format into a {@code RecordQuery}.
+     *
+     * @param <I>
+     *         the type of the record identifiers
+     * @param <R>
+     *         the type of the records which are queried
+     * @param filters
+     *         the original Protobuf-based filters to convert
+     * @param format
+     *         the original response format to convert
+     * @param spec
+     *         the specification of the columns stored for the particular record type
+     * @return a new record query with the same semantic as the original filters and response format
+     */
+    public static <I, R extends Message> RecordQuery<I, R>
+    convert(TargetFilters filters, ResponseFormat format, RecordSpec<I, R, ?> spec) {
+        checkNotNull(spec);
+        checkNotNull(filters);
+        checkNotNull(format);
+
+        Class<I> idType = spec.idType();
+        Class<R> recordType = spec.recordType();
+        RecordQueryBuilder<I, R> builder = RecordQuery.newBuilder(idType, recordType);
+
+        identifiers(builder, filters.getIdFilter());
+        filters(builder, spec, filters);
+        fieldMask(builder, format);
+        orderByAndLimit(builder, spec, format);
+
+        return builder.build();
+    }
+
+    /**
+     * Creates a new {@code RecordQuery} based on the given response format and
+     * the storage specification of the queried record.
+     *
+     * <p>The result contains no filters on any record field.
+     *
+     * @param spec
+     *         the specification of the columns stored for the particular record type
+     * @param format
+     *         the original response format
+     * @param <I>
+     *         the type of the record identifiers
+     * @param <R>
+     *         the type of the records which are queried
+     * @return a new record query
+     */
+    public static <I, R extends Message> RecordQuery<I, R>
+    newQuery(RecordSpec<I, R, ?> spec, ResponseFormat format) {
+        checkNotNull(spec);
+        checkNotNull(format);
+
+        Class<I> idType = spec.idType();
+        Class<R> recordType = spec.recordType();
+        RecordQueryBuilder<I, R> builder = RecordQuery.newBuilder(idType, recordType);
+        fieldMask(builder, format);
+        orderByAndLimit(builder, spec, format);
+        return builder.build();
+    }
+
+    private static <I, R extends Message> void
+    filters(RecordQueryBuilder<I, R> builder, RecordSpec<I, R, ?> spec, TargetFilters filters) {
+        if (filters.getFilterCount() > 0) {
```

Comment:
Why do we want to check the count? The generated code uses `List` instance anyway. What if we just inline the `compositeFilters` variable and remove the `if` check? It would make the whole this method not that much tested.

If you wanted to make it more obvious that nothing happens if the list is empty, why not just do it at the beginning of the method and return right away if so?